### PR TITLE
Support propagation and retain Job resource

### LINKED ISSUE #527: failed to create job
**What happened**:
execution controller failed to dispatch job. 

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
```
E0714 14:03:39.426857       1 execution_controller.go:236] Failed to create resource in the given member cluster member2, err is Job.batch "pi" is invalid: [spec.template.metadata.labels[controller-uid]: Invalid value: map[string]string{"controller-uid":"7ade503c-dc0b-416e-8939-e5e1e4bc6b15", "job-name":"pi"}: must be '8b65be30-5875-4c11-89f3-c9f50e8da2e7', spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"controller-uid":"7ade503c-dc0b-416e-8939-e5e1e4bc6b15"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: `selector` not auto-generated]
E0714 14:03:39.426863       1 execution_controller.go:180] Failed to create resource(default/pi) in the given member cluster member2, err is Job.batch "pi" is invalid: [spec.template.metadata.labels[controller-uid]: Invalid value: map[string]string{"controller-uid":"7ade503c-dc0b-416e-8939-e5e1e4bc6b15", "job-name":"pi"}: must be '8b65be30-5875-4c11-89f3-c9f50e8da2e7', spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"controller-uid":"7ade503c-dc0b-416e-8939-e5e1e4bc6b15"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: `selector` not auto-generated]
```

**Anything else we need to know?**:
job template:
```
apiVersion: batch/v1
kind: Job
metadata:
  name: pi
spec:
  template:
    spec:
      containers:
      - name: pi
        image: perl
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
```
propagation policy:
```
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: job-propagation
spec:
  resourceSelectors:
    - apiVersion: batch/v1
      kind: Job
      name: pi
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
```

### PR DESCRIPTION
**Pull request by mrlihanbo**
Signed-off-by: lihanbo <lihanbo2@huawei.com>

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
kube apiserver will adds a selector to a job and labels to its template when `if job.Spec.ManualSelector == nil || *job.Spec.ManualSelector == false` . So we need to delete the `generateSelector` in work before distributed to member cluster.

**Which issue(s) this PR fixes**:
Fixes #527 

**Special notes for your reviewer**:
after mutating:
```
  spec:
    workload:
      manifests:
      - apiVersion: batch/v1
        kind: Job
        metadata:
          labels:
            controller-uid: d548ad7b-a23c-47a8-9618-b7e71613f21b
            job-name: pi
            propagationpolicy.karmada.io/name: job-propagation
            propagationpolicy.karmada.io/namespace: default
            resourcebinding.karmada.io/name: pi-job
            resourcebinding.karmada.io/namespace: default
            work.karmada.io/name: pi-84fc8846fb
            work.karmada.io/namespace: karmada-es-member2
          name: pi
          namespace: default
        spec:
          backoffLimit: 4
          completions: 1
          parallelism: 1
          selector:
            matchLabels: {}
          template:
            metadata:
              creationTimestamp: null
              labels: {}
            spec:
              containers:
              - command:
                - perl
                - -Mbignum=bpi
                - -wle
                - print bpi(2000)
                image: perl
                imagePullPolicy: Always
                name: pi
                resources: {}
                terminationMessagePath: /dev/termination-log
                terminationMessagePolicy: File
              dnsPolicy: ClusterFirst
              restartPolicy: Never
              schedulerName: default-scheduler
              securityContext: {}
              terminationGracePeriodSeconds: 30
```

**Does this PR introduce a user-facing change?**:
"NONE"

### CODE REVIEW

**Comment by pigletfly** [comment]
so can we find a better way to deal with this kind of problem,not hard-coded in karmada?Maybe we can provide a rule to tell karmada which fields should be removed or retained in a object.Otherwise we have to deal with all kinds of possibilities.

**Comment by RainbowMango** [comment]
@pigletfly 
I've been thinking about this too, and we can do it like [custom-health-checks](https://argo-cd.readthedocs.io/en/stable/operator-manual/health/#custom-health-checks).

We can define default retain rules for well-known resources, meanwhile, users can define the rule by themselves to replace the default ones.

How do you think?

**Comment by pigletfly** [comment]
agreed, and also add new rules.How about using cue-lang,tutorial  https://github.com/cue-lang/cue/tree/master/doc/tutorial/kubernetes

**Comment by RainbowMango** [comment]
> agreed, and also add new rules.How about using cue-lang,tutorial https://github.com/cue-lang/cue/tree/master/doc/tutorial/kubernetes

Will investigate the language.

/kind feature
/remove-kind bug

File: pkg/webhook/work/mutating.go [hunk 1]
```diff
@@ -106,4 +111,54 @@ func removeIrrelevantField(workload *unstructured.Unstructured) {
...
+
+		if job.Spec.ManualSelector == nil || !*job.Spec.ManualSelector {
+			return removeGenerateSelectorOfJob(workload)
+		}
+	}
+
+	return nil
+}
+
+func removeGenerateSelectorOfJob(workload *unstructured.Unstructured) error {
```
Comment (RainbowMango) [inline pkg/webhook/work/mutating.go:130]: The `removeGenerateSelectorOfJob` is anti-operation of [generateSelector](https://github.com/kubernetes/kubernetes/blob/5a732dcfe1d4ec0e8ee2871b106605b7f8a69b98/pkg/registry/batch/job/strategy.go#L215:6), we can't track the change when updating kubernetes, (the behavior might be different), so an E2E test is necessary.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**